runtime/connectors/blob/blobdownloader.go

```diff
@@ -3,119 +3,272 @@ package blob
 import (
 	"context"
 	"fmt"
+	"io"
+	"os"
+	"path/filepath"
 
 	"cloud.google.com/go/storage"
 	"github.com/bmatcuk/doublestar/v4"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/connectors"
 	"github.com/rilldata/rill/runtime/pkg/fileutil"
 	"gocloud.dev/blob"
+	"golang.org/x/sync/errgroup"
 )
 
-type FetchConfigs struct {
-	GlobMaxTotalSize      int64
-	GlobMaxObjectsMatched int
-	GlobMaxObjectsListed  int64
-	GlobPageSize          int
+// increasing this limit can increase speed ingestion
+// but may increase bottleneck at duckdb or network/db IO
+// set without any benchamarks
+const _concurrentBlobDownloadLimit = 8
+
+// map of supoprted extensions for partial downloads vs readers
+var _partialDownloadReaders = map[string]string{".parquet": "parquet", ".csv": "csv", ".tsv": "csv", ".txt": "csv", ".parquet.gz": "parquet"}
+
+// implements connector.FileIterator
+type blobIterator struct {
+	ctx        context.Context
+	bucket     *blob.Bucket
+	objects    []*objectWithPlan
+	index      int
+	localFiles []string
+	// all localfiles are created in this dir
+	tempDir string
 }
 
-// downloads file to local paths
-// todo :: return blob handler as iterator
-func FetchFileNames(ctx context.Context, bucket *blob.Bucket, config FetchConfigs, globPattern, bucketPath string) ([]string, error) {
-	validateConfigs(&config)
-	prefix, glob := doublestar.SplitPattern(globPattern)
+// NewIterator returns new instance of blobIterator
+// the iterator keeps list of blob objects eagerly planned as per user's glob pattern and extract policies
+// clients should call close once done to release all resources like closing the bucket
+// the iterator takes responsibility of closing the bucket
+func NewIterator(ctx context.Context, bucket *blob.Bucket, opts Options, globPattern string) (connectors.FileIterator, error) {
+	opts.validate()
 
-	handler := &BlobHandler{
-		prefix: prefix,
-		bucket: bucket,
-		path:   bucketPath,
+	items, err := plan(ctx, bucket, opts, globPattern)
+	if err != nil {
+		bucket.Close()
+		return nil, err
 	}
 
-	if !fileutil.IsGlob(glob) {
-		// glob represent plain object
-		handler.FileNames = []string{globPattern}
-		err := handler.DownloadAll(ctx)
-		if err != nil {
-			return nil, err
-		}
-		return handler.LocalPaths, nil
+	tempDir, err := os.MkdirTemp(os.TempDir(), "blob*ingestion")
```

Comment:
The * in this string will be replaced with a random string. I will add a comments for that as well.

The idea was to keep directory per source ingestion unique. Since we are also deleting this directory on iterator close we don't want multiple source ingestions to store files in same directory. 
Although as of now we do source ingestion one by one but in case we want to change this in future to allow multiple sources to be ingested in parallel this will be hard to remember and fix. The os.TempDir seems to be unique at least in a single process.

Reply:
Sorry, I totally agree with all of the above. My note was just about `os.TempDir()` and `*` being redundant since it's the default behavior of `MkdirTemp`: https://pkg.go.dev/os#MkdirTemp

Reply:
Yes. That makes sense. Sorry I missed the random suffix part in MkdirTemp